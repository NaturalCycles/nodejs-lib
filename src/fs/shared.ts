import * as fs from 'fs-extra'
import { boldWhite } from '../colors'

export interface GenGlobOptions {
  /**
   * @default to `**` (everything, including sub-dirs)
   */
  inputPatterns?: string[]

  /**
   * @default . (cwd)
   */
  baseDir: string

  /**
   * @default . (cwd)
   */
  outputDir: string
}

export function globPrepare(opt: GenGlobOptions): void {
  // Default pattern
  if (!opt.inputPatterns?.length) opt.inputPatterns = ['**']

  // default to cwd
  opt.baseDir ||= '.'
  opt.outputDir ||= '.'

  if (!fs.existsSync(opt.baseDir)) {
    console.log(`kpy: baseDir doesn't exist: ${boldWhite(opt.baseDir)}`)
    return
  }

  fs.ensureDirSync(opt.outputDir)
}
