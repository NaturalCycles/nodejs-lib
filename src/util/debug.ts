import debug from 'debug'
import { IDebugger } from 'debug'
import * as path from 'path'

export function getDebug (filename: string): IDebugger {
  return debug(getDebugNamespaceFromFilename(filename))
}

function getDebugNamespaceFromFilename (filename: string): string {
  // drop last part after dot
  const a = path.basename(filename).split('.')
  a.pop()
  return a.join('.')
}
