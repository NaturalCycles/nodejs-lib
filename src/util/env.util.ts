import type { ValuesOf } from '@naturalcycles/js-lib'
import 'dotenv/config' // ensure .env is read before requiring keys
import * as fs from 'fs-extra'

/**
 * @example
 *
 * const {a, b} = requreEnvKeys(['a', 'b'])
 *
 * Will throw if any of the passed keys is not defined.
 */
export function requireEnvKeys<T extends readonly string[]>(
  ...keys: T
): { [k in ValuesOf<T>]: string } {
  // eslint-disable-next-line unicorn/no-array-reduce
  return keys.reduce((r, k) => {
    const v = process.env[k]
    if (!v) throw new Error(`${k} env variable is required, but missing`)
    r[k] = v
    return r
  }, {} as { [k in ValuesOf<T>]: string })
}

export function requireFileToExist(filePath: string): void {
  if (!fs.pathExistsSync(filePath)) {
    throw new Error(`Required file should exist: ${filePath}`)
  }
}
