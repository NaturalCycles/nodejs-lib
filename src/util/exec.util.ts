import type { ProcessEnvOptions, SpawnOptions } from 'node:child_process'
import * as cp from 'node:child_process'

export async function execVoidCommand(
  cmd: string,
  args: string[] = [],
  opt: SpawnOptions = {},
): Promise<void> {
  logExec(cmd, args, opt)

  await new Promise<void>(resolve => {
    const p = cp.spawn(cmd, [...args], {
      stdio: 'inherit',
      // shell: true,
      ...opt,
      env: {
        ...process.env,
        ...opt.env,
      },
    })

    p.on('close', code => {
      if (!code) return resolve()
      console.log(`${cmd} exited with code ${code}`)
      process.exit(code)
    })
  })
}

export function execVoidCommandSync(
  cmd: string,
  args: string[] = [],
  opt: SpawnOptions = {},
): void {
  logExec(cmd, args, opt)

  const r = cp.spawnSync(cmd, [...args], {
    encoding: 'utf8',
    stdio: 'inherit',
    // shell: true, // removing shell breaks executing `tsc`
    ...opt,
    env: {
      ...process.env,
      ...opt.env,
    },
  })

  if (r.status) {
    console.log(`${cmd} exited with code ${r.status}`)
    process.exit(r.status)
  }

  if (r.error) {
    console.log(r.error)
    process.exit((r.error as NodeJS.ErrnoException).errno || 1)
  }
}

function logExec(cmd: string, args: string[] = [], opt: ProcessEnvOptions = {}): void {
  const cmdline = [
    ...Object.entries(opt.env || {}).map(([k, v]) => [k, v].join('=')),
    cmd,
    ...args,
  ].join(' ')

  console.log(cmdline)
}
