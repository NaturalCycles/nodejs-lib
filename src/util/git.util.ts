import * as cp from 'node:child_process'
import * as path from 'node:path'
import type { UnixTimestampNumber } from '@naturalcycles/js-lib'
import { grey } from '../colors/colors'

export function getLastGitCommitMsg(): string {
  return execSync('git log -1 --pretty=%B')
}

export function commitMessageToTitleMessage(msg: string): string {
  const firstLine = msg.split('\n')[0]!
  const [preTitle, title] = firstLine.split(': ')
  return title || preTitle!
}

export function gitHasUncommittedChanges(): boolean {
  // git diff-index --quiet HEAD -- || echo "untracked"
  try {
    cp.execSync('git diff-index --quiet HEAD --', {
      encoding: 'utf8',
    })
    return false
  } catch {
    return true
  }
}

/**
 * @returns true if there were changes
 */
export function gitCommitAll(msg: string): boolean {
  // git commit -a -m "style(lint-all): $GIT_MSG" || true
  const cmd = `git commit -a --no-verify -m "${msg}"`
  // const cmd = `git`
  // const args = ['commit', '-a', '--no-verify', '-m', msg]

  console.log(grey(cmd))

  try {
    cp.execSync(cmd, {
      stdio: 'inherit',
    })
    return true
  } catch {
    return false
  }
}

/**
 * @returns true if there are not pushed commits.
 */
export function gitIsAhead(): boolean {
  // ahead=`git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
  const cmd = `git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`

  const stdout = execSync(cmd)

  // console.log(`gitIsAhead: ${stdout}`)
  return Number(stdout) > 0
}

export function gitPull(): void {
  const cmd = 'git pull'
  try {
    cp.execSync(cmd, {
      stdio: 'inherit',
    })
  } catch {}
}

export function gitPush(): void {
  // git push --set-upstream origin $CIRCLE_BRANCH && echo "pushed, exiting" && exit 0
  let cmd = 'git push'

  const branchName = gitCurrentBranchName()

  if (branchName) {
    cmd += ` --set-upstream origin ${branchName}`
  }

  console.log(grey(cmd))

  cp.execSync(cmd, {
    stdio: 'inherit',
  })
}

export function gitCurrentCommitSha(full = false): string {
  const sha = execSync('git rev-parse HEAD')
  return full ? sha : sha.slice(0, 7)
}

export function gitCurrentCommitTimestamp(): UnixTimestampNumber {
  return Number(execSync('git log -1 --format=%ct'))
}

export function gitCurrentBranchName(): string {
  return execSync('git rev-parse --abbrev-ref HEAD')
}

export function gitCurrentRepoName(): string {
  const originUrl = execSync('git config --get remote.origin.url')
  return path.basename(originUrl, '.git')
}

function execSync(cmd: string): string {
  return cp
    .execSync(cmd, {
      encoding: 'utf8',
      // stdio: 'inherit', // no, otherwise we don't get the output returned
    })
    .trim()
}
