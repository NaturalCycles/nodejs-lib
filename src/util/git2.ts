import cp from 'node:child_process'
import path from 'node:path'
import type { UnixTimestampNumber } from '@naturalcycles/js-lib'
import { grey } from '../colors/colors'
import { exec2 } from './exec2'

/**
 * Set of utility functions to work with git.
 */
class Git2 {
  getLastGitCommitMsg(): string {
    return exec2.exec('git log -1 --pretty=%B')
  }

  commitMessageToTitleMessage(msg: string): string {
    const firstLine = msg.split('\n')[0]!
    const [preTitle, title] = firstLine.split(': ')
    return title || preTitle!
  }

  gitHasUncommittedChanges(): boolean {
    // git diff-index --quiet HEAD -- || echo "untracked"
    try {
      cp.execSync('git diff-index --quiet HEAD --', {
        encoding: 'utf8',
      })
      return false
    } catch {
      return true
    }
  }

  /**
   * Returns true if there were changes
   */
  gitCommitAll(msg: string): boolean {
    // git commit -a -m "style(lint-all): $GIT_MSG" || true
    const cmd = `git commit -a --no-verify -m "${msg}"`
    // const cmd = `git`
    // const args = ['commit', '-a', '--no-verify', '-m', msg]
    console.log(grey(cmd))

    try {
      cp.execSync(cmd, {
        stdio: 'inherit',
      })
      return true
    } catch {
      return false
    }
  }

  /**
   * @returns true if there are not pushed commits.
   */
  gitIsAhead(): boolean {
    // ahead=`git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
    const cmd = `git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
    const stdout = exec2.exec(cmd)
    // console.log(`gitIsAhead: ${stdout}`)
    return Number(stdout) > 0
  }

  gitPull(): void {
    const cmd = 'git pull'
    try {
      cp.execSync(cmd, {
        stdio: 'inherit',
      })
    } catch {}
  }

  gitPush(): void {
    // git push --set-upstream origin $CIRCLE_BRANCH && echo "pushed, exiting" && exit 0
    let cmd = 'git push'

    const branchName = this.gitCurrentBranchName()

    if (branchName) {
      cmd += ` --set-upstream origin ${branchName}`
    }

    exec2.spawn(cmd, { logStart: true })
  }

  gitCurrentCommitSha(full = false): string {
    const sha = exec2.exec('git rev-parse HEAD')
    return full ? sha : sha.slice(0, 7)
  }

  gitCurrentCommitTimestamp(): UnixTimestampNumber {
    return Number(exec2.exec('git log -1 --format=%ct'))
  }

  gitCurrentBranchName(): string {
    return exec2.exec('git rev-parse --abbrev-ref HEAD')
  }

  gitCurrentRepoName(): string {
    const originUrl = exec2.exec('git config --get remote.origin.url')
    return path.basename(originUrl, '.git')
  }
}

export const git2 = new Git2()
