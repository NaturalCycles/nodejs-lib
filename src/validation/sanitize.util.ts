import * as sanitize from 'sanitize-html'

export type SanitizeHTMLOptions = sanitize.IOptions

/**
 * Simply a wrapper around `sanitize-html` library.
 *
 * @experimental
 */
export function sanitizeHTML(s: string, opt?: SanitizeHTMLOptions): string {
  return sanitize(s, opt)
}
