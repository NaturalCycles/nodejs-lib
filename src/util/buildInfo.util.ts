import {
  _filterUndefinedValues,
  AnyObject,
  BuildInfo,
  localTimeOrNow,
  UnixTimestampNumber,
} from '@naturalcycles/js-lib'
import { fs2 } from '../fs/fs2'
import {
  gitCurrentBranchName,
  gitCurrentCommitSha,
  gitCurrentCommitTimestamp,
  gitCurrentRepoName,
} from './git.util'

export interface GenerateBuildInfoOptions {
  /**
   * If set - this timestamp will be used, instead of "current time".
   */
  overrideTimestamp?: UnixTimestampNumber
}

export function generateBuildInfo(opt: GenerateBuildInfoOptions = {}): BuildInfo {
  const now = localTimeOrNow(opt.overrideTimestamp)
  const ts = now.unix()

  const rev = gitCurrentCommitSha()
  const branchName = gitCurrentBranchName()
  const repoName = gitCurrentRepoName()
  const tsCommit = gitCurrentCommitTimestamp()

  const ver = [now.toStringCompact(), repoName, branchName, rev].join('_')

  let { APP_ENV: env } = process.env

  if (!env) {
    // Attempt to read `envByBranch` from package.json root
    try {
      if (fs2.pathExists('package.json')) {
        const packageJson = fs2.readJson<AnyObject>('package.json')
        env = packageJson?.['envByBranch']?.[branchName] || packageJson?.['envByBranch']?.['*']
      }
    } catch {}
  }

  return _filterUndefinedValues({
    ts,
    tsCommit,
    repoName,
    branchName,
    rev,
    ver,
    env,
  })
}
