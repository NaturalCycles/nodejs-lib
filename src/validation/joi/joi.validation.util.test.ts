import { arraySchema, objectSchema, stringSchema } from './joi.shared.schemas'
import { JoiValidationError } from './joi.validation.error'
import { getValidationResult, validate, validationErrorToString } from './joi.validation.util'

interface Obj1 {
  a1: string
  a2?: string
}

const obj1Schema = objectSchema({
  a1: stringSchema.min(2).max(5),
  a2: stringSchema.min(2).optional(),
})

interface Obj2 {
  s1?: string
  a: Obj1[]
}

const obj2Schema = objectSchema({
  s1: stringSchema.optional(),
  a: arraySchema.items(obj1Schema),
})

const invalidValues: any[] = [
  undefined,
  null,
  '',
  'abc',
  5,
  () => 'a',
  { a1: 'a' },
  { a1: 'a12345' },
]

const validValues: any[] = [
  {
    a1: 'ff',
  },
  {
    a1: 'ff',
    a2: 'sdfer',
  },
]

test('should fail on invalid values', () => {
  invalidValues.forEach(v => {
    expect(() => validate(v, obj1Schema)).toThrowErrorMatchingSnapshot()
  })
})

test('should pass on valid values', () => {
  validValues.forEach(v => {
    validate(v, obj1Schema)
  })
})

test('should trim strings by default', async () => {
  const v = { a1: ' sdf ' }
  const v2 = validate(v, obj1Schema)

  expect(v2.a1).toBe('sdf')
  expect(v === v2).toBeFalsy() // object should be cloned
})

test('should strip unknown keys', async () => {
  const v = {
    a1: 'ff',
    unk: 'ddd',
  }
  const v2 = validate(v, obj1Schema)

  expect(v2).toEqual({ a1: 'ff' })
  expect(v2.unk).toBeUndefined()
})

test('getValidationResult should still convert', async () => {
  const v = {
    a1: ' ff ', // to be converted
    a2: 'a', // invalid!
  }
  const vr = getValidationResult(v, obj1Schema, 'objName')
  expect(vr.value.a1).toBe('ff')
  expect(vr.error).toBeInstanceOf(JoiValidationError)
  expect(vr.error).toMatchSnapshot()
})

test('getValidationResult valid', async () => {
  const vr = getValidationResult('asd', stringSchema)
  expect(vr.error).toBeUndefined()
})

test('validationErrorToString', async () => {
  expect(validationErrorToString(undefined as any)).toBeUndefined()
})

test('error should contain errorItems', async () => {
  const v = {
    a1: ' ff ', // to be converted
    a2: 'a', // invalid!
  }
  const { error } = getValidationResult(v, obj1Schema, 'objName')
  expect(error!.data).toMatchSnapshot()
})

test('array items with unknown props', async () => {
  const v: Obj2 = {
    a: [
      {
        a1: 'hello',
        unk: 'unk', // unknown property!
      } as Obj1,
    ],
  }

  const { value, error } = getValidationResult(v, obj2Schema)
  // console.log(value)
  expect(error).toBeUndefined()

  // Expect 'unk' to be stripped away, no validation error
  expect(value).toEqual({
    a: [{ a1: 'hello' }],
  })
})

test('array with invalid items', async () => {
  const v: Obj2 = {
    a: [
      {
        a1: 'hello',
      } as Obj1,
      '' as any, // invalid value
    ],
  }

  const { error } = getValidationResult(v, obj2Schema)
  // console.log(value)
  expect(error!.data).toMatchSnapshot()
})

test('array items with invalid props', async () => {
  const v: Obj2 = {
    a: [
      {
        a1: 5, // invalid type
      } as any,
    ],
  }

  const { error } = getValidationResult(v, obj2Schema)
  expect(error!.data).toMatchSnapshot()
})

// optional
// default values

test('long message string', () => {
  const objSchema = arraySchema.items(
    objectSchema({
      a: stringSchema,
    }),
  )

  const longObject = Array(1000).fill({ a: 5 })

  const { error } = getValidationResult(longObject, objSchema)
  // console.log(error!.message, error!.message.length)
  expect(error!.message).toMatchSnapshot()
})
