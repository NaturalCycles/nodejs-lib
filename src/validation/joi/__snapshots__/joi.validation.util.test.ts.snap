// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array items with invalid props 1`] = `
Object {
  "joiValidationErrorItems": Array [
    Object {
      "context": Object {
        "key": "a1",
        "label": "a1",
        "value": 5,
      },
      "message": "\\"a1\\" must be a string",
      "path": Array [
        "a",
        0,
        "a1",
      ],
      "type": "string.base",
    },
  ],
}
`;

exports[`array with invalid items 1`] = `
Object {
  "joiValidationErrorItems": Array [
    Object {
      "context": Object {
        "key": 1,
        "label": 1,
        "value": "",
      },
      "message": "\\"1\\" must be an object",
      "path": Array [
        "a",
        1,
      ],
      "type": "object.base",
    },
  ],
}
`;

exports[`error should contain errorItems 1`] = `
Object {
  "joiValidationErrorItems": Array [
    Object {
      "context": Object {
        "encoding": undefined,
        "key": "a2",
        "label": "a2",
        "limit": 2,
        "value": "a",
      },
      "message": "\\"a2\\" length must be at least 2 characters long",
      "path": Array [
        "a2",
      ],
      "type": "string.min",
    },
  ],
}
`;

exports[`getValidationResult should still convert 1`] = `
[JoiValidationError: objName
{
  "a1": " ff ",
  "a2" [31m[1][0m: "a"
}
[31m
[1] "a2" length must be at least 2 characters long[0m]
`;

exports[`should fail on invalid values 1`] = `"\\"value\\" is required"`;

exports[`should fail on invalid values 2`] = `
"{
  [41m\\"value\\"[0m[31m [1]: -- missing --[0m
}
[31m
[1] \\"value\\" must be an object[0m"
`;

exports[`should fail on invalid values 3`] = `"\\"value\\" must be an object"`;

exports[`should fail on invalid values 4`] = `"\\"value\\" must be an object"`;

exports[`should fail on invalid values 5`] = `"\\"value\\" must be an object"`;

exports[`should fail on invalid values 6`] = `"\\"value\\" must be an object"`;

exports[`should fail on invalid values 7`] = `
"{
  \\"a1\\" [31m[1][0m: \\"a\\"
}
[31m
[1] \\"a1\\" length must be at least 2 characters long[0m"
`;

exports[`should fail on invalid values 8`] = `
"{
  \\"a1\\" [31m[1][0m: \\"a12345\\"
}
[31m
[1] \\"a1\\" length must be less than or equal to 5 characters long[0m"
`;
